File: ThreeFluids/getData.c
/* Title: Janus Drops
# Author: Vatsal Sanjay
# vatsalsanjay@gmail.com
# Physics of Fluids
*/
#include "navier-stokes/centered.h"

scalar f1[], f2[], f3[];
char filename[80];
int nx,ny;
double xmin, ymin, xmax, ymax;
scalar * list = NULL;

event init(t = 0)
{
  restore (file = filename);
  FILE * fp = ferr;
  nx++;
  ny++;
  double Deltax = 0.999999*(xmax-xmin)/(nx - 1);
  double Deltay = 0.999999*(ymax-ymin)/(ny - 1);
  int len = list_len(list);
  double ** field = (double **) matrix_new (nx, ny, len*sizeof(double));
  for (int i = 0; i < nx; i++) {
    double x = Deltax*i + xmin;
    for (int j = 0; j < ny; j++) {
      double y = Deltay*j + ymin;
      int k = 0;
      for (scalar s in list){
        field[i][len*j + k++] = interpolate (s, x, y);
      }
    }
  }

  for (int i = 0; i < nx; i++) {
    double x = Deltax*i + xmin;
    for (int j = 0; j < ny; j++) {
      double y = Deltay*j + ymin;
      fprintf (fp, "%g %g", x, y);
      int k = 0;
      for (scalar s in list){
        fprintf (fp, " %g", field[i][len*j + k++]);
      }
      fputc ('\n', fp);
    }
  }
  fflush (fp);
  // fclose (fp);
  matrix_free (field);
}

int main(int a, char const *arguments[])
{
  sprintf (filename, "%s", arguments[1]);
  xmin = atof(arguments[2]);
  xmax = atof(arguments[3]);
  ymin = atof(arguments[4]);
  ymax = atof(arguments[5]);
  nx = atoi(arguments[6]);
  ny = atoi(arguments[7]);
  list = list_add (list, f1);
  list = list_add (list, f2);
  list = list_add (list, f3);
  // list = list_add (list, u.x);
  // list = list_add (list, u.y);
  run();
}

File: ThreeFluids/getData.m
%%
% Author: Vatsal Sanjay
% vatsalsanjay@gmail.com
% Physics of Fluids
function [X, Y, f1, f2, f3] = getData(datafile, xmin, xmax, ymin, ymax, nx, ny)
disp('Using ./getData to extract data from Basilisk C!')    
tic
exe = sprintf('!./getData %s %s %s %s %s %s %s', datafile, num2str(xmin), num2str(xmax), num2str(ymin), num2str(ymax), num2str(nx), num2str(ny));
ll=evalc(exe); 
bolo=textscan(ll,'%f %f %f %f %f\n'); 
toc
X = reshape(bolo{1},[nx+1,ny+1]);
Y = reshape(bolo{2},[nx+1,ny+1]);
f1 = reshape(bolo{3},[nx+1,ny+1]);
f2 = reshape(bolo{4},[nx+1,ny+1]);
f3 = reshape(bolo{5},[nx+1,ny+1]);
end
